Change data_home default to avoid Flexvolume plugin conflict in GKE

twilfong opened the pull request:
The Scalr agent K8s Helm chart creates a DaemonSet in the worker template that makes use of a hostPath directory. Unfortunately, the default value for this is "/home/kubernetes/flexvolume/agent-k8s", which is a directory that the GKE distribution of Kubernetes uses as its  Flexvolume plugin directory.

GKE changes the default [Flexvolume](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-storage/flexvolume.md) plugin directory from /var/lib/kubelet/volumeplugins to /home/kubernetes/flexvolume, in its Kubelet configuration. (Flexvolume is deprecated but still supported.) If this directory exists, Kubelet automatically scans it for new custom volume driver plugins, which causes (non-critical) errors to be constantly logged by the kubelet on every node in the cluster where this chart is installed.

This directory should be changed to something that no service running on the host should expect to be used for any other purpose. This is a simple one line change to do this, but a longer-term fix might be to move away from using a hostPath directly.

twilfong committed:
Change data_home default to avoid Flexvolume plugin conflict in GKE

The current "/home/kubernetes/flexvolume/agent-k8s" default is a directory that the GKE distribution of Kubernetes uses as its  Flexvolume plugin directory. 

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-storage/flexvolume.md

mermoldy commented:
Hi @twilfong, thank you for pointing this out. We wasn't aware that using flexvolume was causing issues in kubelet services. The problem with GKE's hostPath is that it's currently the only directory with the executable permissions required for tofu/terraform to execute its plugins. Other directories on the root node disk like `/home/scalr` lack execution permissions, leading to issues during tf init:
```
│ Error: failed to read provider configuration schema for registry.terraform.io/hashicorp/null: failed to instantiate provider "registry.terraform.io/hashicorp/null" to obtain schema: fork/exec .terraform/providers/registry.terraform.io/hashicorp/null/3.2.1/linux_amd64/terraform-provider-null_v3.2.1_x5: permission denied
```

A more robust solution for GKE is to use external per-node data stores like https://cloud.google.com/kubernetes-engine/docs/concepts/local-ssd. The `data_dir` can then be mounted to `/mnt/disks/ssd0/`. However, this isn't something that can be included in the default chart configuration.

Using flexvolume is more of a workaround than a permanent solution, and moving away from hostPath usage is definitely a long-term goal.

gabrielrinaldi commented:
I am having an issue related to this as well, I'm running it on EKS, but I've tried a few different paths without success. Also I agree with @mermoldy that this should not be a long term solution.

twilfong commented:
I have tested this on our GKE clusters, by changing the value of agent.data_home to "/home/scalr/agent-k8s" in our helm chart installation, and did run into this error. As you suggest, this is a volume that is mounted with `noexec`, and /home/kubernetes/flexvolume is not mounted with `noexec`. However, in the interest of not using a directory reserved for other purposes, and also to avoid the Kubelet error, I would suggest using a different directory mounted without `noexec` which shouldn't cause any kubelet issues. Doing just a bit more searching, I think that "/home/kubernetes/bin/scalr/agent-k8s" should work. The only potential issue with "/home/kubernetes/bin" is that it is intended for official K8s tools, but I still think it is a better compromise than using the flexvolume directory.

twilfong committed:
Change data_home to /home/kubernetes/bin/scalr/agent-k8s

Unlike "/home", the "/home/kubernetes/bin" directory should not be mounted with noexec, so this should work. This is tested with GKE, but not with other K8s providers.

twilfong commented:
@gabrielrinaldi: We've decided to use "/home/kubernetes/bin/scalr/agent-k8s" (as in my recent commit to the PR). In GKE this is mounted on a volume without the `noexec` option and works for us.

```
$ df /home/kubernetes/bin
Filesystem     1K-blocks     Used Available Use% Mounted on
/dev/sda1       98831908 36577436  62238088  38% /home/kubernetes/bin
$ findmnt /home/kubernetes/bin
TARGET               SOURCE                          FSTYPE OPTIONS
/home/kubernetes/bin /dev/sda1[/home/kubernetes/bin] ext4   rw,nosuid,nodev,relatime,commit=30
```

twilfong commented:
We are able to change the "agent.data_home" value without it being a problem for us, but I think the larger community using this helm chart ought to be aware of the issue that the default value causes for the kubelet. At a minimum this causes a large increase in log volume (and thus increased cost). I'll switch this to an issue instead of a PR since it is apparently not a trivial change to make.

mermoldy commented:
This issue is resolved in the alternative charts - see the more detailed explanation in the related issue: https://github.com/Scalr/agent-helm/issues/33#issuecomment-4374765278
